미션 제출합니다.

owenoweny opened the pull request.

owenoweny committed:
feat: 움직임 기능과 테스트
- Car: 자동차 위치 및 이름, 비교 가능 여부
- MoveChecker: 자동차 움직임 가능 여부 판단
- RandomGenerator: 0-9 사이 난수 생성

owenoweny committed:
feat: 게임 진행 기능 구현

- Game.java: 게임 진행 로직 구현
- Cars.java: Car 일급 컬렉션 클래스 구현
- 그 외 검증 로직 구현

owenoweny committed:
feat: 게임 진행 상태 및 결과 출력

- Cars.java: 우승자 판별
- Game.java -> RacingCarGame.java
- RacingCarGame.java: 시도 회차별 상태 구문
- RacingCarApplication.java: 게임 진행 로직 구현

owenoweny committed:
refactor: 게임 진행 로직과 뷰를 분리

- ConsoleView.java: 각종 인풋과 결과 프롬프트 출력
- RacingCarApplication.java: 코드 스타일을 컨벤션에 맞게 수정

owenoweny committed:
refactor: 패키지 구조 수정

owenoweny committed:
feat: 문자열 입력 파싱 기능 구현

owenoweny committed:
refactor: 파서 구조 리팩토링

- Parser.java: 파서가 상태를 가지게 변경

owenoweny committed:
feat: 덧셈 기능 구현

owenoweny committed:
feat: 계산기 실행 기능 구현

- Parser 생성자 접근 제어자 수정

owenoweny committed:
feat: 계산기 파서에서 정규표현식 예약어를 필터링

owenoweny committed:
feat: 계산기 파서에서 음수 입력을 필터링

owenoweny committed:
docs: 기능 목록 수정

owenoweny committed:
refactor: 패키지 구조 수정, 코드 스타일 통일

owenoweny committed:
fix: CalculatorApp - 입력 문자열을 입력받아 합 정수를 반환하는 형태로 메소드 수정

seokhong commented on README.md:
```diff
@@ -1,9 +1,38 @@
+
+# 레이싱 게임
+
+## 기능 목록
```
기능 목록을 작성할 때 마크다운의 체크리스트 문법을 활용해보면 어떨까요?
단계별로 진행하면서 체크 표시도 하면, 어디까지 진행했는지 업무 진행 과정과 속도를 한 번에 볼 수 있다는 장점도 있습니다. :)

seokhong commented on src/main/java/calculator/domains/Parser.java:
```diff
@@ -0,0 +1,57 @@
+
+    private final List<Integer> numbers;
+
+
```
불필요한 빈 줄 하나가 더 들어간 것으로 보이네요!

seokhong commented on src/main/java/calculator/domains/Parser.java:
```diff
@@ -0,0 +1,57 @@
+        }
+        Matcher m = defaultInputFormatPattern.matcher(input);
+        if (!m.find()) {
+            throw new RuntimeException("");
```
예외 상황에 대한 내용도 메시지로 전달해보면 어떨까요?

seokhong commented on src/main/java/calculator/domains/Parser.java:
```diff
@@ -0,0 +1,57 @@
+        return new Parser(m.group(1), m.group(2));
+    }
+
+    private static void validateRegex(String input) {
```
꼼꼼한 밸리데이션 좋습니다. 👍

seokhong commented on src/main/java/calculator/domains/Parser.java:
```diff
@@ -0,0 +1,57 @@
+    private static final Pattern defaultInputFormatPattern = Pattern.compile("//(.)\n(.*)");
+    private static final Pattern forbiddenDelimiterPattern = Pattern.compile("\\.|\\^|\\$|\\*|\\+|\\?|\\|");
+
+    private final List<Integer> numbers;
```
`Parser`를 생성할 때 `numbers`를 만들고, `find` 에서는 이 값을 그대로 반환만 하게 되네요.
이 객체를 사용하는 입장에서는 `Parser` 보다는 `Numbers` 라는 이름이 좀 더 이해하기 쉽지 않을까요? 객체가 값을 파싱하는 기능을 제공하지는 않는 것으로 보입니다. `Numbers`라는 이름으로 변경한다면, 객체를 생성하는 과정에서 파싱하는 로직이 복잡한데요. 이 기능을 분리하는 것도 고민해 볼 수 있을 것 같아요. 또한, `Numbers`라는 객체라면 `sum` 과 같은 메서드를 제공하고, 합계를 반환하는 기능을 제공해줄 수도 있을 것 같네요!

owenoweny replied:
@seokhong 피드백 감사합니다!! 클래스 이름이 역할을 제대로 나타내지 못한다고 생각했는데, 피드백 주신 부분 바탕으로 리팩토링을 했습니다. - [98277d5](https://github.com/next-step/java-racingcar-kakao/pull/5/commits/98277d5e7bd293f8331885eb8ad19565d4cb5e78)
`Parser`를 파싱 기능만을 갖고 있는 유틸 클래스로 수정했고, `Numbers` 클래스가 계산 대상이 되는 숫자에 대한 일급 컬랙션의 역할을 하도록 했습니다. 그리고 이 과정에서 다음과 같은 고민들을 했습니다.

1. **`Parser`를 유틸 클래스로 유지하는 것에 대해 확신이 없습니다.** `Parser`가 static 메소드를 제공하는 유틸성 클래스가 되니까, Parser 자체가 상태를 갖지 않고 메소드들 사이에서 인자로 많은 정보를 주고 받는 것 같습니다. 다만 의미 상 파싱은 유틸의 특성을 많이 갖고 있는 것 같아서 이와 같이 구현했는데, 관련해서 의견 주시면 감사드리겠습니다.
2. **`Numbers`와 `Parser`, `CalculatorApp` 사이의 의존관계에서 고민이 많습니다.** 현재는 `Numbers`의 팩토리 메소드에서 정규식을 입력받고, `Parser`에 의존해서 파싱한 후 객체를 생성하도록 했습니다. 다른 대안으로는, `CalculatorApp`에서 콘솔로 입력받은 값을 `List<Integer>`로 파싱해서 `Numbers`에게 넘겨주고, `Numbers`는 단순히 List를 입력받아 객체를 초기화하는 역할만 하도록 할 수 있을 것 같습니다. `CalculatorApp`이 많은 도메인 객체들에 의존하지 않는 편이 좋을 것 같다고 생각했는데, 애매한 것 같습니다. 리뷰어께서는 어떻게 생각하시는 지 궁금합니다.

seokhong replied:
1. 요환님께서는 어떤 기준으로 유틸리티 클래스 사용을 결정하시나요? 메서드에서 인자의 개수가 많지 않다면, 적절한 기능 구현을 위한 인자의 전달은 괜찮다고 생각합니다. 말씀하신 것처럼 상태를 가지고 있고, 인자의 개수가 적은 것이 바람직한 경우도 있지만, 필요한 경우 상태가 있더라도 인자로 전달하는 것이 더 적절하거나 코드를 이해하는데 도움이 되는 경우도 있습니다. 먼저 이 클래스가 유틸리티 클래스로 구현하는게 적절한지, 혹은 `Expression`과 같은 클래스로 만들고 값을 전달한 후, `numbers()`와 같은 메서드를 호출하면 `Numbers` 객체를 만드는 방법은 어떨지 등을 비교해보시는 것도 좋겠습니다.
2. `CalculatorApp`이 많은 도메인 객체에 의존하지 않는 것이 어떤 면에서 좋다고 생각하셨나요? 적절한 의존성의 개수나, 방향 등에 대해 어떻게 생각하시는지 궁금합니다. 저는 클래스나 패키지간의 의존성의 방향이 단방향이고, 코드를 이해하는데 방해가 될 정도로  의존하는 대상이 많은 것이 아니라면 괜찮다고 생각합니다.

seokhong commented on src/main/java/racinggame/RacingCarApplication.java:
```diff
@@ -0,0 +1,27 @@
+
+    public void run() {
+        ConsoleView.printCarNamesInputPrompt();
+        RacingCarGame racingCarGame = RacingCarGame.of(ConsoleView.getCarNamesInput());
```
자동차 이름을 입력받기 위해서는 `ConsoleView.printCarNamesInputPrompt`와 `ConsoleView.getCarNamesInput`을 순서대로 호출해야 하는데요. 이는 `ConsoleView` 외부에서 `ConsoleView` 내부의 구현을 자세히 알아야 한다는 의미가 됩니다.
자동차 이름을 입력받기 위해 외부에서는 `ConsoleView.getCarNamesInput`만 호출하고, 세부 구현은 `ConsoleView`가 알아서 하도록 변경해보면 어떨까요? 여러 public 메서드를 제공하면서 해당 메서드를 순서대로 사용하길 기대하는 건 지양하는 것이 좋습니다.

owenoweny replied:
@seokhong 콘솔 출력과 입력이 분리되어야하는 기능이라고 생각해서 메소드를 분리했었는데, 말씀해주신대로 `ConsoleView`에 의존하는 클라이언트 코드 입장에서 이들을 순서대로 사용해야한다는 추가적인 이해가 필요하다는 단점이 매우 큰 것 같습니다. 피드백 감사합니다!! - [3c87b54](https://github.com/next-step/java-racingcar-kakao/pull/5/commits/3c87b54e88f7bd2026ed3896405a1373927dc6c6)

seokhong commented on src/main/java/racinggame/RacingCarApplication.java:
```diff
@@ -0,0 +1,27 @@
+        int trialInput = ConsoleView.getTrialInput();
+
+        ConsoleView.printResultPrompt();
+        for (int reps = 0; reps < trialInput; reps++) {
```
1. `trialInput`이 0 이하인 경우 어떻게 될까요?
2. 간단한 용어라도 `reps`와 같이 약어를 사용하시지는 않는 것을 추천합니다. 코드의 양이 많아지면 IDE의 문자열 검색 기능도 많이 사용하실텐데요. 이때, 약어는 제대로 검색이 안 되는 경우가 있습니다. 또한, 코드를 읽는 사람이 이 약어의 원래 의미를 생각하기 위해 한 번 더 고민해야 하는 불편함이 있습니다. 구체적인 약어 사용 규칙이 있지 않다면, 항상 원래 단어를 그대로 사용하시는 것을 추천합니다.

owenoweny replied:
reps 정도면 이미 그 의미가 충분히 합의된 단어라고 생각했었는데, 말씀하신대로 더 확실한 규칙이 있어야할 것 같습니다. 또 IDE의 문자열 검색 기능은 고려하지 못한 부분이었습니다. 감사합니다!!

seokhong commented on src/main/java/racinggame/domains/Car.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    @Override
+    public int compareTo(Car o) {
```
1. `Comparable`을 구현하신 것은 좋은데, 사용하는 곳이 테스트 코드밖에 없는 것 같아요.
2. `Car`의 `compareTo` 구현을 안 봤을 때, 처음 보는 개발자는 `compareTo`가 위치를 비교한다고 생각하기 어렵지 않을까요? 명시적으로 `compareOffset` 과 같은 메서드를 제공해주는 건 어떨까요?

owenoweny replied:
@seokhong Comparable은 Cars에서 승자 계산을 위해 sort() 메소드를 이용하기 위해 구현했습니다. 다만 말씀해주신대로, Car 객체 자체의 비교를 offset만으로 처리하는 것이 적합하지 않다고 생각합니다. 2번에서 말씀해주신 부분은 고려하지 못했네요. 피드백 감사합니다. - [c31bb68](https://github.com/next-step/java-racingcar-kakao/pull/5/commits/c31bb685cdee74a4135a175f77ab375ceff74c94) 관련 리팩토링 commit입니다.

seokhong commented on src/main/java/racinggame/domains/Car.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
equals, hashCode 오버라이드 좋습니다. 👍

seokhong commented on src/main/java/racinggame/domains/Cars.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    private void validate(List<Car> cars) {
+        Set<Car> carSet = new HashSet<>(cars);
```
`Car`의 `hashCode` 구현을 봤을 때, 같은 이름을 가지고 있더라도 `offset`이 다르면 다른 객체로 구분할 것으로 보이네요!

owenoweny replied:
객체의 동등성을 판단하는 기준을 고민하지 않은 것 같습니다. 감사합니다!

seokhong replied:
equals, hashCode 구현을 `owner`만 사용하도록 변경하신 것은, 이름이 일치한다면 동일한 객체여야 한다는 제약 조건하에 구현하신 것이라고 이해하면 될까요? 혹시 현재의 코드가 잘 동작하기 위한 수정이라면, 이름 중복을 검사하는 로직을 수정하는 것이 더 적절해 보이기도 합니다.

seokhong commented on src/main/java/racinggame/domains/Cars.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    public List<Car> winners() {
+        cars.sort(Comparator.reverseOrder());
```
객체를 정렬하는 과정은 O(n log n)의 시간 복잡도를 가질 것으로 보이네요.
전체 자동차 목록을 확인하여 가장 멀리 간 차의 위치를 확인한 후, 다시 전체 목록을 확인해서 해당 위치에 있는 차만 가지고 오면 O(n)의 복잡도로 문제를 해결할 수 있지 않을까요? :)

seokhong commented on src/main/java/racinggame/domains/Cars.java:
```diff
@@ -0,0 +1,36 @@
+        cars.sort(Comparator.reverseOrder());
+        return cars
+                .stream()
+                .filter(e -> cars.get(0).offset() <= e.offset())
```
`e.offset()`으로 값을 꺼내서 비교하기보다는, `e.at(cars.get(0).offset())`과 같이 값을 전달해주고 해당 위치에 있는지 물어보는 형태로 구현해보면 어떨까요?

seokhong commented on src/main/java/racinggame/domains/MoveChecker.java:
```diff
@@ -0,0 +1,11 @@
+
+public final class MoveChecker {
+    private MoveChecker() {
+        throw new AssertionError("정적 메서드만 보유한 객체입니다.");
```
`AssertionError`는 검증 조건을 위배할 때 발생시켜야 할 에러로 보이네요.
`InstantiationException` 등의 다른 예외를 사용해보면 어떨까요?

seokhong commented on src/main/java/racinggame/domains/MoveChecker.java:
```diff
@@ -0,0 +1,11 @@
+    }
+
+    public static boolean check(int randomNumber) {
+        return randomNumber >= 4;
```
4라는 값은 의미를 가지는 숫자이니, 상수로 변경하고 이름을 부여해보면 어떨까요?

seokhong commented on src/main/java/racinggame/domains/RacingCarGame.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    public void showStatus() {
+        cars
```
32행처럼 `cars.list()`를 한 줄에 써도 괜찮을 것으로 보이네요. `stream()`도 기능을 수행하지는 않으므로, `cars.list().stream()`과 같이 한 줄로 쓰는 경우도 많은 것 같습니다. :)

seokhong commented on src/main/java/racinggame/utils/RandomGenerator.java:
```diff
@@ -0,0 +1,15 @@
+    }
+
+    public int generate() {
+        return random.nextInt(10);
```
10도 상수로 분리할 수 있겠네요.

seokhong commented on src/test/java/racinggame/utils/RandomGeneratorTest.java:
```diff
@@ -0,0 +1,19 @@
+@DisplayName("랜덤 수를 생성 결과 테스트")
+public class RandomGeneratorTest {
+    @Test
+    @DisplayName("0-9 사이 랜덤 값 발생 100,000회 확인")
```
난수 생성을 테스트하신 이유가 있을까요? :)

owenoweny replied:
@seokhong TDD를 처음 시도하는 상황에서 페어와 "어디까지 테스트해야하는지"를 설정하면서 우선 테스트를 추가해봤습니다. 기능을 검증하는 기능은 없지만, 단위 테스트가 해당 클래스의 기능에 대한 명세로써의 역할도 할 수 있을 것이라고 생각했습니다. 
그래서 사실 이 테스트의 당위성에 대해서는 저희도 확신이 없었습니다. 피드백 주셔서 감사합니다~~

seokhong replied:
테스트를 작성할 때는 이 테스트를 통해 무엇을 검증하고 싶은지, 어떤 변경으로부터 코드를 보호하고 싶은지를 고려해보셔도 도움이 될 것 같습니다. 예를 들어, 난수가 잘 생성되는지 테스트를 하고 싶은 것이라면 100,000회 난수를 생성했을 때 0-9 사이의 다른 값이 적절한 횟수로 생성되는지 검증이 필요하겠죠.
이 테스트에 대해서는 이런 질문도 떠올랐습니다. 난수를 생성하는 기능은 자바 라이브러리의 기능인데 테스트가 필요할까요? 만약 기능이 변경되어 0-5 사이의 값을 생성하도록 코드가 변경된다면 이 테스트는 여전히 성공하지만, 코드의 변경을 알지는 못하는 것 아닐까요? 반대로, 1-10 사이의 값을 생성하도록 코드가 변경되었을 때, (매우 낮은 확률이지만) 이 테스트가 성공하는 일도 발생하지는 않을까요? :)

seokhong reviewed (changes requested):
요환님 안녕하세요.
자동차 경주 미션을 함께 할 리뷰어 김석홍입니다. 만나서 반갑습니다!

코멘트 남긴 내용 확인 부탁드립니다.
코멘트 관련하여 궁금하신 점은 PR에 코멘트 또는 슬랙으로 DM 보내주시면 확인하는대로 회신 드리겠습니다. 🙂

owenoweny committed:
fix: 예외 처리 추가

- 해당 예외를 더 잘 나타내는 예외로 수정
- 예외 메세지 추가
- 레이싱 카 시도 횟수 입력 예외 처리 추가

owenoweny committed:
refactor: 상수 분리 및 코드 스타일 통일

owenoweny committed:
refactor: 상수 분리

owenoweny committed:
test: 불필요한 테스트 삭제

owenoweny committed:
refactor: ConsoleView 메소드 리팩토링

- 차 이름, 반복 횟수 반환 메소드 역할을 수정

owenoweny committed:
refactor: 입력 파싱 로직 리팩토링

- Parser: 파싱 기능만을 갖고 있는 유틸 클래스로 수정
- Numbers: 계산 대상이 되는 숫자의 정보 저장과 계산 역할

owenoweny committed:
refactor: Car 객체 비교 관련 로직 리팩토링

- offset 비교를 위한 compareOffset(), at() 메소드 구현
- Cars.winners() -> 시간 복잡도를 줄이기 위해 로직 수정
- Car가 Comparable을 구현하지 않도록 수정

owenoweny commented:
@seokhong 커멘트해주신 사항 바탕으로 기능 수정, 리팩토링했습니다. 수정 사항과 질문은 커밋 로그와 남겨주신 코멘트에 남겼습니다. 답변 주시면 감사드리겠습니다.

 페어프로그래밍과 PR 리뷰 진행 방식에 대한 이해가 부족해서 리뷰와 관련된 정보를 많이 드리지 못한 것 같습니다. 리뷰 정말 감사합니다!!

seokhong commented on src/main/java/racinggame/ConsoleView.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    public static void printResult(RacingCarGame racingCarGame) {
+        System.out.println(racingCarGame
```
출력 결과를 `System.out.println` 내부에서 만들면 코드의 양은 줄어들지만, 출력 결과가 무엇을 나타내는지는 한 눈에 알기 어려울 수도 있어 보입니다. :)
코드의 양이 조금 늘어나더라도, 지역 변수를 활용해서 `우승자 목록`임을 나타낼 수 있는 이름으로 표현해본다면, 코드를 읽을 때 더 이해하기 쉬울 것 같아요.

seokhong commented on src/main/java/racinggame/domains/Car.java:
```diff
@@ -0,0 +1,60 @@
+import java.util.Objects;
+
+public final class Car {
+    public static final int MAX_CAR_NAME_LENGTH = 5;
```
내부에서만 사용하는 상수는 private 접근 제어자로 관리해보면 어떨까요?
또한 상수와 필드 사이에는 빈 줄을 추가하면 코드를 읽을 때 도움이 됩니다. :)

seokhong commented on src/main/java/racinggame/domains/Car.java:
```diff
@@ -0,0 +1,60 @@
+        return Objects.hash(owner);
+    }
+
+    public int compareOffset(Car c2) {
```
비교 대상이 인자로 들어올 때 `other` 와 같은 이름도 많이 사용하는 것 같아요!

seokhong commented on src/main/java/racinggame/domains/Cars.java:
```diff
@@ -0,0 +1,40 @@
+import java.util.stream.Collectors;
+
+public final class Cars {
+    private final List<Car> cars;
```
필드와 클래스의 이름이 같으면, 아마도 자동차 목록을 가져올 때 아래와 비슷한 코드를 사용하게 될텐데요.
```java
cars.getCars();
```

원시값을 포장할 때 발생할 수 있는 일입니다.
`Optional`이나 `Integer`와 같이 값을 포장하는 클래스를 보시면 `value`, `values`와 같은 이름을 사용하는 것도 괜찮아 보입니다.
포장한 값을 꺼낼 때는 `values()`, `get()`과 같은 이름을 사용하는 방법도 있겠네요.

seokhong commented on src/test/java/calculator/domains/NumbersTest.java:
```diff
@@ -0,0 +1,5 @@
+package calculator.domains;
+
+class NumbersTest {
```
테스트 코드가 없는 테스트는 삭제해도 괜찮을 것 같아요!

seokhong reviewed (approved):
요환님, 1단계 미션 코멘트 잘 반영하셨습니다. 👍 
1단계 PR은 머지하겠습니다.

질문 남겨주신 부분에 대한 제 생각, 그리고 다음 단계 미션 진행하실 때 고민해보시면 좋을 것 같은 내용을 코멘트로 남겼습니다. 참고 부탁드립니다.
요환님의 생각이 궁금한 점도 질문을 남겼는데요. 다음 단계 리뷰 요청하실 때, 관련된 내용 코멘트로 남겨주시면 감사합니다.
2단계 미션에서 뵙겠습니다. 🙂
